### PR TITLE
Fix cleanup of IPv6 BPF map

### PR DESCRIPTION
**Pull request by sargun**
There was a bug on how the key was derived for the IPv6 HTB classid
mapping BPF map.

### CODE REVIEW

File: vpc/tool/container2/setup_container_linux.go [hunk 1]
```diff
@@ -745,25 +727,59 @@ func deleteFromIPv6BPFMap(ctx context.Context, assignment *vpcapi.AssignIPRespon
 	defer unix.Close(int(fd))
 	runtime.KeepAlive(path)
 
+	key := ipv6key(assignment)
```
Comment (solarkennedy) [inline vpc/tool/container2/setup_container_linux.go:730]: is the `buf` even used anymore in this function?

File: vpc/tool/container2/setup_container_linux.go [hunk 2]
```diff
@@ -745,25 +727,59 @@ func deleteFromIPv6BPFMap(ctx context.Context, assignment *vpcapi.AssignIPRespon
...
 		logger.G(ctx).WithError(errors.Wrapf(err, "Unable to delete element for map with file descriptor %d", fd)).Error()
-		err2 := errors.Wrap(err, "Unable to delete element for ipv4 map")
+		err2 := errors.Wrap(err, "Unable to delete element for ipv6 map")
 		return err2
 	}
 	runtime.KeepAlive(uba)
 	return nil
 }
 
+func ipv6key(assignment *vpcapi.AssignIPResponseV3) []byte {
```
Comment (solarkennedy) [inline vpc/tool/container2/setup_container_linux.go:754]: good stuff, making these functions

File: vpc/tool/container2/setup_container_linux.go (hunk 1 above)
Comment (MyUmmaGumma) [inline vpc/tool/container2/setup_container_linux.go:730]: Should we be checking if key is nil - we seem to be accessing key[0] next.
(Same with the v6key below)

**Comment by solarkennedy** [comment]
Since the ipv6 key is changing (but I guess wasn't working in the first place), are you worried about leftover ipv6 keys on existing agents, after this new code is deployed mutably? (or are you thinking they will get auto-remediated like they are today when the map fills up anyway)

File: vpc/tool/container2/setup_container_linux.go [hunk 3]
```diff
@@ -745,25 +727,59 @@ func deleteFromIPv6BPFMap(ctx context.Context, assignment *vpcapi.AssignIPRespon
...
 	)
-	if err != unix.ENOENT && err != 0 {
+	if err == unix.ENOENT {
+		err := errors.Wrap(err, "Unable to delete element for ipv6 map (notfound)")
+		logger.G(ctx).WithError(err).WithField("ip", assignment.Ipv6Address.Address.Address).WithField("classid", assignment.ClassId).WithField("vlanid", assignment.VlanId).Error("Element not found")
+		return err
+	} else if err != 0 {
 		logger.G(ctx).WithError(errors.Wrapf(err, "Unable to delete element for map with file descriptor %d", fd)).Error()
-		err2 := errors.Wrap(err, "Unable to delete element for ipv4 map")
+		err2 := errors.Wrap(err, "Unable to delete element for ipv6 map")
```
Comment (solarkennedy) [inline vpc/tool/container2/setup_container_linux.go:747]: Do we need the err2 pattern? Can you just return `errors.wrap` ?

File: vpc/tool/container2/setup_container_linux.go [hunk 4]
```diff
@@ -745,25 +727,59 @@ func deleteFromIPv6BPFMap(ctx context.Context, assignment *vpcapi.AssignIPRespon
...
+		return nil
+	}
+	buf := make([]byte, 8)
+
+	binary.LittleEndian.PutUint16(buf, uint16(assignment.VlanId))
+	ip := net.ParseIP(assignment.Ipv4Address.Address.Address).To4()
+	if len(ip) != 4 {
+		panic("Length of IP is not 4")
+	}
+	copy(buf[4:], ip)
```
Comment (MyUmmaGumma) [inline vpc/tool/container2/setup_container_linux.go:779]: Out of curiosity - What do we do with the first 4 bytes of this buffer eventually?

**Review by MyUmmaGumma — APPROVED** [review]

File: vpc/tool/container2/setup_container_linux.go [hunk 5]
```diff
@@ -694,17 +671,22 @@ func deleteFromIPv4BPFMap(ctx context.Context, assignment *vpcapi.AssignIPRespon
 	defer unix.Close(int(fd))
 	runtime.KeepAlive(path)
```
Comment (solarkennedy) [inline vpc/tool/container2/setup_container_linux.go:672]: On a second pass, if you wanted to, I think you could DRY out the raw bpf map stuff with a function like:

    deleteKeyFromBPFMap(mapPath string, key []byte)

that would do the fd BPF_OBJ_GET, uba, BPF_MAP_DELETE_ELEM, and return just the error.

and make `"/sys/fs/bpf//tc/globals/ipv4_map\000"` consts 

Maybe. We are talking 2 copies of the code here, and my goto for refactoring is 3 copies. Up to you.

**Review by solarkennedy — APPROVED** [review]

File: vpc/tool/container2/setup_container_linux.go (hunk 4 above)
Comment (sargun) [reply vpc/tool/container2/setup_container_linux.go:779]: VLAN ID / sg

File: vpc/tool/container2/setup_container_linux.go (hunk 3 above)
Comment (sargun) [reply vpc/tool/container2/setup_container_linux.go:747]: because err2 is an "error" whereas err is an syscall.Errno.

**Pull request merged**